platform/Atomic.h

```diff
@@ -0,0 +1,1108 @@
+                             (n) > sizeof(uint8_t)  ? sizeof(uint16_t)   \
+                                                    : sizeof(uint8_t))
+
+template<size_t N>
+struct atomic_size_traits
+{
+    static bool is_lock_free() { return false; };
+};
+
+template<>
+struct atomic_size_traits<sizeof(uint8_t)>
+{
+    typedef uint8_t storage_type;
+    static bool is_lock_free() { return MBED_EXCLUSIVE_ACCESS; }
+};
+
+template<>
+struct atomic_size_traits<sizeof(uint16_t)>
+{
+    typedef uint16_t storage_type;
+    static bool is_lock_free() { return MBED_EXCLUSIVE_ACCESS; }
+};
+
+template<>
+struct atomic_size_traits<sizeof(uint32_t)>
+{
+    typedef uint32_t storage_type;
+    static bool is_lock_free() { return MBED_EXCLUSIVE_ACCESS; }
+};
+
+template<>
+struct atomic_size_traits<sizeof(uint64_t)>
+{
+    typedef uint64_t storage_type;
+    static bool is_lock_free() { return false; }
+};
+
+template<typename T>
+struct atomic_traits : public atomic_size_traits<MBED_ATOMIC_SIZE(sizeof(T))>
+{
+};
+
+namespace mbed {
+
+/** Atomic template
+ *
+ * `mbed::Atomic<T>` is intended to work as per C++14 `std::atomic<T>`. `T` must be a
+ * _TriviallyCopyable_ type.
+ * - All standard methods of `std::atomic` are supplied
+ * - Operations are guaranteed atomic with respect to interrupts, and the
+ *   operations can be used in interrupts - `std::atomic` implementations don't specify this.
+ * - Implementation is optimised for uniprocessor use (no DMB instructions),
+ *   unlike typical `std::atomic` implementations.
+ * - Lock-free versions (LDREX/STREX) are used for user types if small enough and available,
+ *   otherwise critical sections are used.
+ * - If used with large objects, interrupt latency may be impacted.
+ * - `ATOMIC_VAR_INIT` is not supplied. Valid initialisation forms are
+ *       Atomic<int> foo;         // zero initialized if static or thread-local, else value indeterminate
+ *       atomic_init(&foo, 2);    // initialise a default-initialised variable (once only, not atomic)
+ *       Atomic<int> foo(2);      // value initialisation
+ *       Atomic<int> foo = { 2 }; // C++11 or later only - also legal C11 with _Atomic int
+ *       Atomic<int> foo = 2;     // C++17 or later only (legal C11 with _Atomic int)
+ * - `MBED_ATOMIC_XXX_LOCK_FREE` replaces `ATOMIC_XXX_LOCK_FREE` - "locking" forms
+ *   take a critical section, non-locking do not.
+ * - For `bool`, integer types and pointers, storage is compatible with the
+ *   plain types. If necessary, they can be substituted as plain types for C
+ *   compatibility in headers, and accessed using core_util_atomic functions. Eg:
+ *      struct foo {
+ *      #ifdef __cplusplus
+ *         mbed::atomic_uint32_t counter; // Use C++ templates from C++ code
+ *      #else
+ *         uint32_t counter;  // Could use core_util_atomic_xxx_u32 from C code, or just have this for structure layout.
+ *      #endif
+ *      };
+ */
```

Comment:
Part 2 of above comment.
```suggestion
@endcode
 */
```